graph-gateway/src/network/subgraph_client.rs

```diff
@@ -0,0 +1,171 @@
+    #[serde(rename_all = "camelCase")]
+    pub struct SubgraphDeployment {
+        #[serde(rename = "ipfsHash")]
+        pub id: DeploymentId,
+        pub manifest: Manifest,
+        #[serde(rename = "indexerAllocations")]
+        pub allocations: Vec<Allocation>,
+        #[serde(default)]
+        pub transferred_to_l2: bool,
+    }
+
+    #[serde_as]
+    #[derive(Debug, Clone, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct Allocation {
+        pub id: Address,
+        #[serde_as(as = "serde_with::DisplayFromStr")]
+        pub allocated_tokens: u128,
+        pub indexer: Indexer,
+    }
+
+    #[serde_as]
+    #[derive(Debug, Clone, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct Indexer {
+        pub id: Address,
+        pub url: Option<String>,
+        #[serde_as(as = "serde_with::DisplayFromStr")]
+        pub staked_tokens: u128,
+    }
+}
+
+/// The Graph network subgraph client.
+pub struct Client {
+    client: subgraph_client::Client,
+    l2_transfer_support: bool,
+}
+
+impl Client {
+    /// Creates a new [`Client`] instance.
+    pub fn new(client: subgraph_client::Client, l2_transfer_support: bool) -> Self {
+        Self {
+            client,
+            l2_transfer_support,
+        }
+    }
+
+    /// Fetch the list of subgraphs (and deployments) from the network subgraph.
+    ///
+    /// Performs a paginated query to fetch the latest state of the network subgraph:
+    /// > Get all subgraphs ordered by `id` in ascending order, filtering out subgraphs with no
+    /// > versions and with all its `versions` ordered by `version` in ascending order, and all
+    /// > its `indexerAllocations` ordered by `allocatedTokens` in descending order; laying the
+    /// > largest allocation first.
+    #[allow(clippy::obfuscated_if_else)]
+    pub async fn fetch(&self) -> anyhow::Result<Vec<types::Subgraph>> {
+        // The following query's response must fulfill the following assumptions:
+        // - Perform a paginated query based on the subgraph's ID. All subgraphs are ordered by ID.
+        //   The next query must start from the last subgraph ID fetched in the previous query.
+        // - Get all subgraphs with at least one version.
+        // - All the subgraph's versions are ordered by `version` in descending order. We assume
+        //   that the newest deployment version is the one with the highest version number.
+        // - All the subgraph's `indexerAllocations` are ordered by `allocatedTokens` in descending
+        //   order; the largest allocation must be the first one.
+        //
+        // ref#: 9936786a-e286-45f3-9190-8409d8389e88
+        let query = format!(
+            r#"
+            subgraphs(
+                block: $block
+                orderBy: id, orderDirection: asc
+                first: $first
+                where: {{
+                    id_gt: $last
+                    entityVersion: 2
+                    versionCount_gt: 0
```

Comment:
Shouldn't this be `versionCount_gt: 1` to meet the requirements above?
> Get all subgraphs with at least one version.

Reply:
This means `count > 0`, which would be equivalent to `count >= 1`.

Fro clarity, I can change it with `versionCount_gte: 1` (count >= 1).

Reply:
Addressed in [15d3252](https://github.com/edgeandnode/gateway/pull/785/commits/15d3252ecedab95bf24ad9d820d0eac648d18305)